### PR TITLE
Save config maps into disk

### PR DESCRIPTION
**Pull request by pjbgf**
Save config maps into disk whilst blocking file traversal at different components.
This may require some improvements around testing, specially around E2E.

### CODE REVIEW

File: controllers/profile/profile.go [hunk 1]
```diff
@@ -38,7 +42,23 @@ const (
...
-	errGetProfile = "cannot get profile"
+	errGetProfile                = "cannot get profile"
+	errConfigMapNil              = "config map cannot be nil"
+	errConfigMapWithoutNamespace = "config map must have a namespace"
+	errConfigMapWithoutName      = "config map must have a name"
+	errSavingProfile             = "cannot save profile"
+	errCreatingOperatorDir       = "cannot create operator directory"
+
+	seccompOperatorSuffix string      = "operator"
+	filePermissionMode    os.FileMode = 0600
```
Comment (saschagrunert) [inline controllers/profile/profile.go:53]:
```suggestion
	filePermissionMode    os.FileMode = 0o600
```

File: controllers/profile/profile.go [hunk 2]
```diff
@@ -38,7 +42,23 @@ const (
...
+
+	// MkdirAll won't create a directory if it does not have the execute bit.
+	// https://github.com/golang/go/issues/22323#issuecomment-340568811
+	dirPermissionMode os.FileMode = 0700
```
Comment (saschagrunert) [inline controllers/profile/profile.go:57]:
```suggestion
	dirPermissionMode os.FileMode = 0o700
```

File: controllers/profile/profile.go [hunk 3]
```diff
@@ -38,7 +42,23 @@ const (
...
+)
+
+var (
+	kubeletSeccompRootPath string = "/var/lib/kubelet/seccomp"
```
Comment (saschagrunert) [inline controllers/profile/profile.go:61]: This can be `const` I guess.

File: controllers/profile/profile.go [hunk 4]
```diff
@@ -88,10 +108,60 @@ func (r *Reconciler) Reconcile(req reconcile.Request) (reconcile.Result, error)
 		return reconcile.Result{}, errors.Wrap(ignoreNotFound(err), errGetProfile)
 	}
 
+	for name, contents := range profile.Data {
+		profilePath, err := getProfilePath(name, profile)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+
+		saveProfileOnDisk(profilePath, contents)
```
Comment (saschagrunert) [inline controllers/profile/profile.go:117]: I think we should handle the returned error.

File: controllers/profile/profile.go (hunk 3 above)
Comment (pjbgf) [reply controllers/profile/profile.go:61]: Initially I created it as a `const`, but then changed it to make it testable - allowing tests to fiddle file/folder permissions without needing to run the tests as root. 

Assuming we will initially only support the operator to run as a `DaemonSet` inside the cluster (as opposed to running directly into the host nodes), this becomes less of a link to the actual seccomp root path on the host, but rather what is the mapped path inside the container. So I was wondering whether is there a better way to tackle this - any suggestions?

File: controllers/profile/profile.go (hunk 4 above)
Comment (pjbgf) [reply controllers/profile/profile.go]: good spot, I completely overlooked this before - it should now be handled now.

File: controllers/profile/profile.go (hunk 3 above)
Comment (saschagrunert) [reply controllers/profile/profile.go:61]: Okay let's leave it like it is for now. Later on we can put this value into a configuration struct to ensure easier testing.

File: controllers/profile/profile.go [hunk 5]
```diff
@@ -88,10 +108,64 @@ func (r *Reconciler) Reconcile(req reconcile.Request) (reconcile.Result, error)
...
+			return errors.Wrap(err, errCreatingOperatorDir)
+		}
+		if os.IsNotExist(err) {
+			if err := os.MkdirAll(dirPath, dirPermissionMode); err != nil {
+				return errors.Wrap(err, errSavingProfile)
+			}
+		}
+	}
+
+	if err := ioutil.WriteFile(fileName, []byte(contents), filePermissionMode); err != nil {
```
Comment (hasheddan) [inline controllers/profile/profile.go:141]: Do we want to validate that the contents are actually a valid profile configuration?

Comment (pjbgf) [reply controllers/profile/profile.go:141]: That is a very good point. I was already planning a follow-up PR with that.

**Comment by saschagrunert** [comment]
Cool, merged that one for now. :)

**Pull request merged**